### ISSUE TITLE
INF issue during training

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#65]
I am getting inf results only during training.  I do not know which loss item is causing this.  This only began occurring when I added the span task. 

it is not the validation loss but the training loss encountering this issue. There is no transparency in the training loss as to which task is causing this

Below is an example resulting from your last attempt. Training loss members are still obscure.  

treamed 493000 items from 'train', collected 493000 valid samples... Streamed 494000 items from 'train', collected 494000 valid samples... Streamed 495000 items from 'train', collected 495000 valid samples... Streamed 496000 items from 'train', collected 496000 valid samples... Streamed 497000 items from 'train', collected 497000 valid samples... Streamed 498000 items from 'train', collected 498000 valid samples... Streamed 499000 items from 'train', collected 499000 valid samples... Streamed 500000 items from 'train', collected 500000 valid samples... Reached max_samples (500000) while streaming 'train'. Collected 500000 samples from 'train' split via streaming. Total collected samples (500000) reached/exceeded max_samples (500000). Stopping C4 loading. Total samples collected via streaming for allenai/c4: 500000 Returning dataset splits: train=400000, val=50000, test=50000 Tokenizing train split... Processed train for Levenshtein: 400000 text items. Tokenizing validation split... Processed validation for Levenshtein: 50000 text items. Tokenizing test split... Processed test for Levenshtein: 50000 text items. train Combined multi-task dataset: 400000 examples Limiting validation for multi-task to approx 48 documents from 50000 for faster evaluation. validation Combined multi-task dataset: 48 examples Limiting test for multi-task to approx 48 documents from 50000 for faster evaluation. test Combined multi-task dataset: 48 examples

--- Preparing train_dataloader --- Dataset object: <combined_dataset.CombinedMultiTaskDataset object at 0x2ac02983ae10> len(dataset): 400000 dataset.seq_len: 1024 --- Preparing train_dataloader (using StrictRatioBatchSampler) --- /usr/local/lib/python3.11/site-packages/torch/utils/data/sampler.py:68: UserWarning: data_source argument is not used and will be removed in 2.2.0.You may still have custom implementation that utilizes it. warnings.warn( batch_sampler: StrictRatioBatchSampler instance num_workers: 2 pin_memory: True train dataloader created: 25000 batches --- train_dataloader preparation complete ---

--- Preparing validation_dataloader --- Dataset object: <combined_dataset.CombinedMultiTaskDataset object at 0x2ac1207706d0> len(dataset): 48 dataset.seq_len: 1024 batch_size: 16 shuffle: False num_workers: 2 collate_fn: None (using default) pin_memory: True validation dataloader created: 3 batches --- validation_dataloader preparation complete ---

--- Preparing test_dataloader --- Dataset object: <combined_dataset.CombinedMultiTaskDataset object at 0x2ac12140e250> len(dataset): 48 dataset.seq_len: 1024 batch_size: 16 shuffle: False num_workers: 2 collate_fn: None (using default) pin_memory: True test dataloader created: 3 batches --- test_dataloader preparation complete --- DEBUG: In entry.py, data_builder.create_dataloaders call completed. Vocab size from data_builder: 259 (UTF-8 bytes potentially extended for Levenshtein CLS) train: 25000 batches of size 16 validation: 3 batches of size 16 test: 3 batches of size 16

=== Data Sample ===

=== Setting up Trainer === Running in non-distributed mode on device: cuda Trainer initialized on device: cuda. Model parameters: 810,041,349

=== Initial Evaluation === Val LM Comp: 0.0000 Val RankReg Aux: 0.0000 Val NSP: 0.0000 Val SpanSelect: 0.0000 Val LM Comp: 6.1928 Val RankReg Aux: 0.1178 Val NSP: 1.5931 Val SpanSelect: 4.6864 Initial training loss (combined): inf Initial validation loss (combined): 6.8797

=== Testing Causal vs Non-Causal Attention === Warning: Train dataloader is empty. Skipping test_causal_attention.

=== Starting Training === Starting training for 1 epochs... Training batches per epoch (Rank 0 view): 25000 Validation batches (Rank 0 view): 3 Calculated steps_per_epoch: 25000 Scheduler re-initialized with T_0 = 2500 based on steps_per_epoch. Val LM Comp: 6.1928 Val RankReg Aux: 0.1072 Val NSP: 1.5988 Val SpanSelect: 5.7971 Initial validation loss (Rank 0): 6.9860 Epoch 1 completed (Rank 0): Avg Loss: nan, Time: 0.29s Checkpoint saved by Rank 0: checkpoints/checkpoint_step_4.pt Training metrics saved (Rank 0): checkpoints/training_metrics.pt

=== Training Completed === Val LM Comp: 0.0000 Val RankReg Aux: 0.0000 Val NSP: 0.0000 Val SpanSelect: 0.0000 Val LM Comp: 6.1928 Val RankReg Aux: 0.1113 Val NSP: 1.5791 Val SpanSelect: 2.1852 Final training loss (combined): inf Final validation loss (combined): 6.6249 Training loss improvement (combined): nan Validation loss improvement (combined): 0.2548

=== Plotting Results === /root/lab/KernelDev/train_loop.py:994: UserWarning: No artists with labels found to put in legend. Note that artists whose label start with an underscore are ignored when legend() is called with no argument. ax_flat[plot_idx].set_xlabel('Step'); ax_flat[plot_idx].set_ylabel('Loss'); ax_flat[plot_idx].grid(True, alpha=0.3); ax_flat[plot_idx].legend() Figure(1500x1500)

=== Testing Text Generation === Generating sample text... Generated text (prompt: '{'task': 'lm', 'prompt': ''}'): �s,�p']O}�q�` �'E�[MASK]�݌N���t ��/��α�}Ӂ����q��

Generated text (prompt: '{'task': 'lm', 'prompt': 'The'}'): '0[CLS]The�o��}��H,�h�$ �����������jUg䥄q�&�p�r'

Generated text (prompt: '{'task': 'lm', 'prompt': 'In a world where'}'): '0[CLS]In a world where�A�j�F䐐�����{FQ-�\������g��X�7�![SEP]7���'1-� �X�'

Generated text (prompt: '{'task': 'unshuffle', 'prompt': 'robot a about story short me Tell'}'): '0[CLS]robot a about story short me Tell$� ��@�Kt �[ [SEP]������ �O�F��XA������,�2 ��O'

Generated text (prompt: '{'task': 'lm', 'type': 'CLS_Prefix_ON', 'prompt': '[CLS] What is the capital of France?'}'): '0[CLS]What is the capital of France?֖�O����r�� �)<ʬ��� �n�n��:��{Lm/ ��s+-[SEP]��.ܝ-'

Generated text (prompt: '{'task': 'lm', 'type': 'CLS_Prefix_OFF', 'prompt': '[CLS] What is the capital of France?'}'): �an�/��s�m�p����Ш ��B��-J`�GS��y[MASK]"Ss:n�'

Generated text (prompt: '{'task': 'lm', 'type': 'CLS_Prefix_ON', 'prompt': '[CLS] Tell me a short story about a robot.'}'): '0[CLS]Tell me a short story about a robot./��+C��a B��rN/ ܝ��1/ /X/����LN��y�r/��}n�6q�'

Generated text (prompt: '{'task': 'lm', 'type': 'CLS_Prefix_OFF', 'prompt': '[CLS] Tell me a short story about a robot.'}'): '0[CLS]Tell me a short story about a robot.�O��Y �����K''�0�ڶ�}�}�/�a/\�b�� x�yƝ�[��2�'

=== Best Results === Best validation loss: 6.6249 at step 5 Total training steps: 6

=== Training Session Complete ===

Remember you are incapable of actually running this code so do not attempt to.

Below is how I have been and expect to continue running this code
# !apt-get update -y && apt-get install -y software-properties-common
# !add-apt-repository ppa:git-core/ppa
# !apt-get install -y git
# !apt-get install -y build-essential
# !pip install torch ninja datasets matplotlib triton
# # # # # # # # # # !pip install -U datasets huggingface_hub fsspec
!rm -rf KernelDev
!git clone --branch fix/inf-loss-in-rank-regression https://github.com/Eupham/KernelDev.git
# # !python KernelDev/original_kernel.py
# # !python KernelDev/test_incoherent_default.py
!python KernelDev/entry.py --nproc_per_node=1 --config KernelDev/config.yaml --epochs 1 --precision bf16
